pubsub/google/cloud/pubsub_v1/publisher/batch/mp.py

```diff
@@ -0,0 +1,398 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+
+import copy
+import multiprocessing
+import queue
+import time
+import uuid
+
+import six
+
+from google.cloud.pubsub_v1 import types
+from google.cloud.pubsub_v1.publisher import exceptions
+from google.cloud.pubsub_v1.publisher.batch import base
+
+
+class Batch(base.BaseBatch):
+    """A batch of messages.
+
+    The batch is the internal group of messages which are either awaiting
+    publication or currently in-flight.
+
+    A batch is automatically created by the PublisherClient when the first
+    message to be published is received; subsequent messages are added to
+    that batch until the process of actual publishing _starts_.
+
+    Once this occurs, any new messages sent to :meth:`publish` open a new
+    batch.
+
+    If you are using this library, you most likely do not need to instantiate
+    batch objects directly; they will be created for you. If you want to
+    change the actual batching settings, see the ``batching`` argument on
+    :class:`~.pubsub_v1.PublisherClient`.
+
+    Any properties or methods on this class which are not defined in
+    :class:`~.pubsub_v1.publisher.batch.BaseBatch` should be considered
+    implementation details.
+
+    Args:
+        client (~.pubsub_v1.PublisherClient): The publisher client used to
+            create this batch.
+        topic (str): The topic. The format for this is
+            ``projects/{project}/topics/{topic}``.
+        settings (~.pubsub_v1.types.Batching): The settings for batch
+            publishing. These should be considered immutable once the batch
+            has been opened.
+        autocommit (bool): Whether to autocommit the batch when the time
+            has elapsed. Defaults to True unless ``settings.max_latency`` is
+            inf.
+    """
+    def __init__(self, client, topic, settings, autocommit=True):
+        self._client = client
+        self._manager = multiprocessing.Manager()
+
+        # Create a namespace that is owned by the client manager; this
+        # is necessary to be able to have these values be communicable between
+        # processes.
+        self._shared = self.manager.Namespace()
+        self._shared.futures = self.manager.list()
+        self._shared.messages = self.manager.list()
+        self._shared.message_ids = self.manager.dict()
+        self._shared.settings = settings
+        self._shared.status = self.Status.ACCEPTING_MESSAGES
+        self._shared.topic = topic
+
+        # This is purely internal tracking.
+        self._process = None
+        self._size = 0
+
+        # Continually monitor the thread until it is time to commit the
+        # batch, or the batch is explicitly committed.
+        if autocommit and self.settings.max_latency < float('inf'):
+            self._process = multiprocessing.Process(target=self.monitor)
```

Comment:
I don't think this makes sense. We should start the timer when the first message is added to the batch, not when the batch object itself is created.